libbeat/processors/annotate/kubernetes/podwatcher.go

```diff
@@ -0,0 +1,184 @@
+	syncPeriod          time.Duration
+	podQueue            chan *corev1.Pod
+	nodeFilter          k8s.Option
+	lastResourceVersion string
+	ctx                 context.Context
+	stop                context.CancelFunc
+	annotationCache     annotationCache
+}
+
+type annotationCache struct {
+	sync.Mutex
+	annotations map[string]common.MapStr
+	pods        map[string]*corev1.Pod // pod uid -> Pod
+}
+
+// NewPodWatcher initializes the watcher client to provide a local state of
+// pods from the cluster (filtered to the given host)
+func NewPodWatcher(kubeClient *k8s.Client, syncPeriod time.Duration, host string) *PodWatcher {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &PodWatcher{
+		kubeClient:          kubeClient,
+		syncPeriod:          syncPeriod,
+		podQueue:            make(chan *corev1.Pod, 10),
+		nodeFilter:          k8s.CustomOption("spec.nodeName", host),
+		lastResourceVersion: "0",
+		ctx:                 ctx,
+		stop:                cancel,
+		annotationCache: annotationCache{
+			annotations: make(map[string]common.MapStr),
+			pods:        make(map[string]*corev1.Pod),
+		},
+	}
+}
+
+func (p *PodWatcher) syncPods() error {
+	logp.Info("kubernetes", "Performing a pod sync")
+	pods, err := p.kubeClient.CoreV1().ListPods(
+		p.ctx,
+		"",
+		p.nodeFilter,
+		k8s.ResourceVersion(p.lastResourceVersion))
+
+	if err != nil {
+		return err
+	}
+
+	for _, pod := range pods.Items {
+		p.podQueue <- pod
+	}
+
+	// Store last version
+	p.lastResourceVersion = pods.Metadata.GetResourceVersion()
+
+	logp.Info("kubernetes", "Pod sync done")
+	return nil
+}
+
+func (p *PodWatcher) watchPods() error {
+	logp.Info("kubernetes", "Watching API for pod events")
+	watcher, err := p.kubeClient.CoreV1().WatchPods(p.ctx, "", p.nodeFilter)
+	if err != nil {
+		return err
+	}
+	for {
+		_, pod, err := watcher.Next()
+		if err != nil {
+			return err
+		}
+
+		p.podQueue <- pod
+	}
+}
+
+func (p *PodWatcher) Run() bool {
+	// Start pod processing worker:
+	go p.worker()
+
+	// Make sure that events don't flow into the annotator before informer is fully set up
+	// Sync initial state:
+	synced := make(chan struct{})
+	go func() {
+		p.syncPods()
+		close(synced)
+	}()
+
+	select {
+	case <-time.After(timeout):
+		p.Stop()
+		return false
+	case <-synced:
+		// Watch for new changes
+		go p.watchPods()
+		return true
+	}
+}
+
+func (p *PodWatcher) addPod(pod *corev1.Pod) {
+	metadata := Indexing.GetMetadata(pod)
+	p.annotationCache.Lock()
+	defer p.annotationCache.Unlock()
+
+	p.annotationCache.pods[pod.Metadata.GetUid()] = pod
+
+	for _, m := range metadata {
+		p.annotationCache.annotations[m.Index] = m.Data
+	}
+}
+
+func (p *PodWatcher) updatePod(pod *corev1.Pod) {
+	oldPod := p.GetPod(pod.Metadata.GetUid())
+	if oldPod.Metadata.GetResourceVersion() != pod.Metadata.GetResourceVersion() {
+		//Process the new pod changes
+		p.deletePod(oldPod)
+		p.addPod(pod)
+	}
+}
+
+func (p *PodWatcher) deletePod(pod *corev1.Pod) {
```

Comment:
this is the naming convention that kube controllers have for various events that are processed which are:
```
onAdd
onUpdate
onDelete
```

we could rename it to `onPodAdd`, `onPodUpdate` and `onPodDelete`. does that work @ruflin ?